File: CoreTextCoreSwift/CoreTextCoreSwift/CTDisplayView.m
//
//  CTDisplayView.m
//  CoreTextCore
//
//  Created by 韩志峰 on 2021/1/20.
//

#import "CTDisplayView.h"
#import <CoreText/CoreText.h>
#import "CoreTextData.h"
#import "CoreTextImageData.h"
#import "CoreUntil.h"
@implementation CTDisplayView



- (void)drawRect:(CGRect)rect {
   

    [super drawRect:rect];
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetTextMatrix(context, CGAffineTransformIdentity);
    CGContextTranslateCTM(context, 0, self.bounds.size.height);
    CGContextScaleCTM(context, 1.0, -1.0);

    if (self.data) {
        CTFrameDraw(self.data.ctframe, context);
        for (CoreTextImageData * imageData in self.data.imageArray) {
            UIImage * image = [UIImage imageNamed: imageData.name];
            CGContextDrawImage(context, imageData.imagePosition, image.CGImage);
        }
    }
    
}
- (instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupEvents];
    }
    return  self;
}
- (void)setupEvents{
    UIGestureRecognizer *tapGrsture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(userTaped:)];
    [self addGestureRecognizer:tapGrsture];
    self.userInteractionEnabled = YES;
}
- (void)userTaped:(UITapGestureRecognizer *)gesture{
    CGPoint point = [gesture locationInView:self];
    for (CoreTextImageData *imageData in self.data.imageArray) {
        //coreImageData中的坐标系 跟UIKit坐标系翻转
        CGRect imageRect = imageData.imagePosition;
        CGPoint imagePosition = imageRect.origin;
        imagePosition.y = self.bounds.size.height - imageRect.origin.y - imageRect.size.height;
        CGRect rect = CGRectMake(imagePosition.x, imagePosition.y, imageRect.size.width, imageRect.size.height);
        if (CGRectContainsPoint(rect, point)) {
            NSLog(@"picture was clicked");
            break;
        }
        
    }
    CoreTextLinkData *linkData = [CoreUntil touchLinkInView:self atPoint:point data:self.data];
    if (linkData) {
        NSLog(@"%@",linkData.url);
    }
    
}


@end

File: CoreTextCoreSwift_other/CoreTextCoreSwift/CTDisplayView.swift
//
//  CTDisplayView.swift
//  CoreTextCoreSwift
//
//  Created by Jz D on 2022/4/24.
//

import Foundation




class CTDisplayView: UIView{
    
    var data: CoreTextData?
    
    
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        let tapGrsture = UITapGestureRecognizer(target: self, action: #selector(userTaped))
        addGestureRecognizer(tapGrsture)
        isUserInteractionEnabled = true
    }
    
    
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        
        guard let context = UIGraphicsGetCurrentContext() , let d_d = data else { return }
        context.textMatrix = .identity
        
        context.translateBy(x: 0, y: bounds.size.height)
        context.scaleBy(x: 1.0, y: -1.0)
        CTFrameDraw(d_d.ctframe, context);
        for imageData in d_d.imageArray{
            if let image = UIImage(named: imageData.name), let cg = image.cgImage{
                context.draw(cg, in: imageData.imagePosition)
            }
        }
        
        
    }
    
    
    @objc func userTaped(_ gesture: UITapGestureRecognizer){
        let point = gesture.location(in: self)
        guard let d = data else { return }
        for imageData in d.imageArray{
            //coreImageData中的坐标系 跟UIKit坐标系翻转
            let imageRect = imageData.imagePosition
            var imagePosition = imageRect.origin
            imagePosition.y = bounds.size.height - imageRect.origin.y - imageRect.size.height;
            let rect = CGRect(x: imagePosition.x, y: imagePosition.y, width: imageRect.size.width, height: imageRect.size.height)
            if rect.contains(point){
                print("picture was clicked")
                break
            }
            
        }
        let linkData = CoreUntil.touchLink(in: self, at: point, data: d)
        print(linkData.url)
        
        
    }
}
